diff --git "a/G:\\Java-Dev\\Vulkan-Tutorial-Java\\src\\main\\java\\javavulkantutorial\\Ch07ImageViews.java" "b/G:\\Java-Dev\\Vulkan-Tutorial-Java\\src\\main\\java\\javavulkantutorial\\Ch08GraphicsPipeline.java"
index 769cc3d..9d995aa 100644
--- "a/G:\\Java-Dev\\Vulkan-Tutorial-Java\\src\\main\\java\\javavulkantutorial\\Ch07ImageViews.java"
+++ "b/G:\\Java-Dev\\Vulkan-Tutorial-Java\\src\\main\\java\\javavulkantutorial\\Ch08GraphicsPipeline.java"
@@ -23,7 +23,7 @@ import static org.lwjgl.vulkan.KHRSurface.*;
 import static org.lwjgl.vulkan.KHRSwapchain.*;
 import static org.lwjgl.vulkan.VK10.*;
 
-public class Ch07ImageViews {
+public class Ch08GraphicsPipeline {
 
     private static class HelloTriangleApplication {
 
@@ -159,6 +159,7 @@ public class Ch07ImageViews {
             createLogicalDevice();
             createSwapChain();
             createImageViews();
+            createGraphicsPipeline();
         }
 
         private void mainLoop() {
@@ -458,7 +459,7 @@ public class Ch07ImageViews {
                     createInfo.subresourceRange().baseArrayLayer(0);
                     createInfo.subresourceRange().layerCount(1);
 
-                    if (vkCreateImageView(device, createInfo, null, pImageView) != VK_SUCCESS) {
+                    if(vkCreateImageView(device, createInfo, null, pImageView) != VK_SUCCESS) {
                         throw new RuntimeException("Failed to create image views");
                     }
 
@@ -468,6 +469,10 @@ public class Ch07ImageViews {
             }
         }
 
+        private void createGraphicsPipeline() {
+
+        }
+
         private VkSurfaceFormatKHR chooseSwapSurfaceFormat(VkSurfaceFormatKHR.Buffer availableFormats) {
             return availableFormats.stream()
                     .filter(availableFormat -> availableFormat.format() == VK_FORMAT_B8G8R8_UNORM)
